File: src/sp/it/pl/gui/infonode/InfoTable.java
package sp.it.pl.gui.infonode;

import java.util.List;
import java.util.function.BiFunction;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.scene.control.Labeled;
import javafx.scene.control.TableView;
import org.reactfx.Subscription;
import static sp.it.pl.util.text.UtilKt.plural;

/**
 * Provides information about table items and table item selection.
 *
 * @param <E> type of table element
 */
public final class InfoTable<E> implements InfoNode<TableView<E>> {

	/**
	 * Default text factory. Provides texts like: <pre>
	 * 'All: 1 item'
	 * 'Selected: 89 items'
	 * </pre>
	 * Custom implementation change or expand the text with additional
	 * information depending on type of table elements.
	 */
	public static final BiFunction<Boolean,List<?>,String> DEFAULT_TEXT_FACTORY = (all, list) -> {
		String prefix1 = all ? "All: " : "Selected: ";
		int s = list.size();
		return prefix1 + s + " " + plural("item", s);
	};

	/**
	 * The graphical text element
	 */
	public Labeled node;

	/**
	 * Provides text to the node. The first parameters specifies whether selection
	 * is empty, the other is the list of table items if selection is empty or
	 * selected items if nonempty.
	 */
	public BiFunction<? super Boolean,? super List<? extends E>, ? extends String> textFactory = DEFAULT_TEXT_FACTORY;

	private Subscription s;

	/**
	 * Sets the node and listeners to update the text automatically by monitoring
	 * the table items and selection.
	 */
	public InfoTable(Labeled node, TableView<E> t) {
		this(node);
		bind(t);
	}

	public InfoTable(Labeled node) {
		this.node = node;
	}

	@Override
	public void setVisible(boolean v) {
		node.setVisible(v);
	}

	@Override
	public void bind(TableView<E> bindable) {
		unbind();

		ObservableList<E> al = bindable.getItems();
		ObservableList<E> sl = bindable.getSelectionModel().getSelectedItems();
		ListChangeListener<E> l = o -> {
			if (o.next()) {
				updateText(al, sl);
			}
		};

		al.addListener(l);
		sl.addListener(l);
		s = () -> al.removeListener(l);
		s = s.and(() -> sl.removeListener(l));

		updateText(al, sl);
	}

	@Override
	public void unbind() {
		if (s!=null) s.unsubscribe();
		s = null;
	}

	/**
	 * Updates the text of the node using the text factory.
	 *
	 * @param all all items of the table
	 * @param selected selected items of the table
	 */
	public void updateText(List<E> all, List<E> selected) {
		boolean isAll = selected.isEmpty();
		List<E> l = isAll ? all : selected;
		node.setText(textFactory.apply(isAll, l));
	}
}

File: src/sp/it/pl/gui/infonode/InfoTable.kt
package sp.it.pl.gui.infonode

import javafx.scene.control.Labeled
import javafx.scene.control.TableView
import org.reactfx.Subscription
import sp.it.pl.util.reactive.onChange
import sp.it.pl.util.text.plural

/**
 * Provides information about table items and table item selection.
 *
 * @param <E> type of table element
 */
class InfoTable<E>: InfoNode<TableView<E>> {

    /** The graphical text element */
    @JvmField var node: Labeled
    /**
     * Provides text to the node. The first parameters specifies whether selection
     * is empty, the other is the list of table items if selection is empty or
     * selected items if nonempty.
     */
    @JvmField var textFactory: (Boolean, List<*>) -> String
    private var s: Subscription? = null

    constructor(node: Labeled) {
        this.node = node
        this.textFactory = DEFAULT_TEXT_FACTORY
    }

    /**
     * Creates infoTable for the labeled and invokes [bind] with the table
     */
    constructor(node: Labeled, t: TableView<E>): this(node) {
        bind(t)
    }

    override fun setVisible(v: Boolean) {
        node.isVisible = v
    }

    override fun bind(bindable: TableView<E>) {
        unbind()

        val listAll = bindable.items
        val listSelected = bindable.selectionModel.selectedItems
        val handleChange = { updateText(listAll, listSelected) }

        handleChange()
        s = Subscription.multi(
                listAll.onChange(handleChange),
                listSelected.onChange(handleChange)
        )
    }

    override fun unbind() {
        s?.unsubscribe()
        s = null
    }

    /**
     * Updates the text of the node using the text factory.
     *
     * @param allItems all items of the table
     * @param selectedItems selected items of the table
     */
    fun updateText(allItems: List<E>, selectedItems: List<E>) {
        val isAll = selectedItems.isEmpty()
        val items = if (isAll) allItems else selectedItems
        node.text = textFactory(isAll, items)
    }

    companion object {

        /** Default text factory. Provides texts like 'All: 1 item' or 'Selected: 89 items'. */
        @JvmField val DEFAULT_TEXT_FACTORY: (Boolean, List<*>) -> String = { isSelectionEmpty, list ->
            val prefix = if (isSelectionEmpty) "All: " else "Selected: "
            val size = list.size
            "$prefix$size ${"item".plural(size)}"
        }
    }
}